docs/development_best_practices.md

```diff
@@ -0,0 +1,130 @@
+
+* File output operator that writes data to files keeps track of the file lengths in the state. These lengths are checkpointed and restored on failure recovery. On restart, the operator truncates the file to the offset equal to the recovered length. This makes the data in the file same as it was at the time of checkpoint before the failure. The operator now writes the replayed data from the checkpoint in regular fashion as any other data. This ensures no data is lost or duplicated in the file.
+* The JDBC output operator that writes data to a database table writes the data in a window in a single transaction. It also writes the current window id into a meta table along with the data as part of the same transaction. It commits the transaction at the end of the window. When there is an operator failure before the final commit, the state of the database is that it contains the data from the previous fully processed window and its window id since the current window transaction isn’t yet committed. On recovery, the operator reads this window id back from the meta table. It ignores all the replayed windows whose window id is less than or equal to the recovered window id and thus ensures that it does not duplicate data already present in the database. It starts writing data normally again when window id of data becomes greater than recovered window thus ensuring no data is lost.
+
+## Partitioning
+
+Partitioning allows an operation to be scaled to handle more pieces of data than before but with a similar SLA. This is done by creating multiple instances of an operator and distributing the data among them. Input operators can also be partitioned to stream more pieces of data into the application. Apex platform provides a lot of flexibility and options for partitioning. Partitioning can happen once at startup or can be dynamically changed anytime while the application is running, and it can be done in a stateless or stateful way by distributing state from the old partitions to new partitions.
+
+In Apex, the responsibility for partitioning is shared among different entities. These are
+ 
+1. A **partitioner** that specifies *how* to partition the operator, specifically it takes an old set of partitions and creates a new set of partitions. At the start of the application the old set has one partition and the partitioner can return more than one partitions to start the application with multiple partitions. The partitioner can use any custom JAVA logic to determine the number of new partitions, set their initial state as a brand new state or by deriving it from the state of the old partitions, and specify how the data gets distributed between the partitions.
```

Comment:
custom JAVA logic => custom logic
between the partitions => among the new partitions.
Worth adding: The old and new set of partitions may intersect, i.e. some old partitions may be preserved unchanged in the new set if desired.